### PR TITLE
[NT-0] feat: Copying asset collections from one space to another

### PR DESCRIPTION
**Pull request by MAG-SamBirley**
This PR introduces the ability for a client application to invoke the duplication of a set of asset collections from one source space to another destination space. It requires that the set of source asset collections all come from the same space.

Some things to be aware of about this change:
* From a CSP API point of view, It changes the notion that an asset collection can belong to _many_ spaces to just _one_. This is to align with the copying API, and also to strengthen this concept in general.
* It also changes the behaviour of one of the constructors for `csp::common::Array` - when an array is constructed from a raw ptr, it now performs a per-element copy (via the = operator) rather than memcpying the block of memory - this resolves issues around inadvertent shared-memory ownership

### CODE REVIEW

File: Library/src/Systems/Assets/AssetCollection.cpp [hunk 1]
```diff
@@ -93,12 +94,15 @@ void PrototypeDtoToAssetCollection(const chs::PrototypeDto& Dto, csp::systems::A
...
+		auto& GroupIds = Dto.GetGroupIds();
+		if (GroupIds.size() > 0)
+		{
+			AssetCollection.SpaceId = GroupIds[0];
+		}
+		else
 		{
-			AssetCollection.SpaceIds[i] = GroupIds[i];
+			CSP_LOG_MSG(csp::systems::LogLevel::Error,
+						"Encountered an asset collection that did not belong to any space whilst processing a response from services.");
```
Comment (MAG-SamBirley) [inline Library/src/Systems/Assets/AssetCollection.cpp:105]: It is valid for an asset collection to not be associated with a space, so this should not log an error

Comment (MAG-SamBirley) [reply Library/src/Systems/Assets/AssetCollection.cpp]: Addressed in latest. Please can I get a re-review @MAG-AW @MAG-AdamThorn?

File: Library/include/CSP/Common/Array.h [hunk 2]
```diff
@@ -74,7 +74,15 @@ template <typename T> class CSP_API Array
...
-			memcpy(ObjectArray, Buffer, Size * sizeof(T));
+
+			// Perform the copy per-element via the assignment operator, rather than memcpy.
+			// This is to avoid introducing inadvertent shared memory ownership between elements in different arrays.
+			// We do not, for example, want strings in an array to hold the same pointers to
+			// memory as another array containing strings that it was constructed from.
+			for (size_t i = 0; i < Size; i++)
+			{
+				ObjectArray[i] = Buffer[i];
+			}
```
Comment (MAG-MichaelK) [inline Library/include/CSP/Common/Array.h:85]: The purpose of this constructor was to be able to "move" a raw C-style array into a `csp::common::Array`. The raw array would then be freed by the caller.

File: Library/include/CSP/Systems/Assets/AssetCollection.h [hunk 3]
```diff
@@ -71,16 +71,37 @@ class CSP_API AssetCollection
...
+
+	// The unique identifier of the POI this asset collection relates to. Empty if it does not relate to a POI.
 	csp::common::String PointOfInterestId;
+
+	// The unique ID of the asset collection that is a parent to this asset collection. Empty if there is no parent relationship.
 	csp::common::String ParentId;
-	csp::common::Array<csp::common::String> SpaceIds;
+
+	// Where the asset collection belongs to a space, the unique ID of the space that this asset collection belongs to. Empty otherwise.
+	csp::common::String SpaceId;
```
Comment (MAG-MichaelK) [inline Library/include/CSP/Systems/Assets/AssetCollection.h:93]: What's the rationale behind this change? This essentially limits us in terms of sharing assets between spaces, and requires a unique copy on CHS for each space. I would rather we instead give the user the _option_ to copy the assets instead of referencing the original.

File: Library/include/CSP/Systems/Assets/AssetSystem.h [hunk 4]
```diff
@@ -76,6 +76,18 @@ class CSP_API CSP_NO_DISPOSE AssetSystem : public SystemBase
 	/// @param Callback NullResultCallback : callback when asynchronous task finishes
 	CSP_ASYNC_RESULT void DeleteAssetCollection(const AssetCollection& AssetCollection, NullResultCallback Callback);
 
+	/// @brief Copies an array of asset collections to another space. Note that all source asset collections must belong to the same space.
```
Comment (MAG-MichaelK) [inline Library/include/CSP/Systems/Assets/AssetSystem.h:79]: Is this a CHS limitation?

File: Library/include/CSP/Systems/Assets/AssetSystem.h [hunk 5]
```diff
@@ -76,6 +76,18 @@ class CSP_API CSP_NO_DISPOSE AssetSystem : public SystemBase
...
+	/// @param SourceAssetCollections csp::common::Array<AssetCollection> : The array of asset collections to copy. They must all belong to the same
+	/// space.
+	/// @param DestSpaceId  const csp::common::String : The unique identifier of the space to copy these asset collections to.
+	/// @param CopyAsync const csp::common::Optional<bool> : Whether to instruct the services to perform the copy of the asset collections
+	/// asynchronously.
+	/// @param Callback NullResultCallback : callback when asynchronous task finishes
+	CSP_ASYNC_RESULT void CopyAssetCollectionsToSpace(csp::common::Array<AssetCollection>& SourceAssetCollections,
```
Comment (MAG-MichaelK) [inline Library/include/CSP/Systems/Assets/AssetSystem.h:86]: What is this in aid of? I see this PR has no ticket, so is this for a personal project?

File: Library/src/Systems/Assets/AssetCollection.cpp [hunk 6]
```diff
@@ -93,12 +94,10 @@ void PrototypeDtoToAssetCollection(const chs::PrototypeDto& Dto, csp::systems::A
 
 	if (Dto.HasGroupIds())
 	{
-		auto& GroupIds			 = Dto.GetGroupIds();
-		AssetCollection.SpaceIds = csp::common::Array<csp::common::String>(GroupIds.size());
-
-		for (int i = 0; i < GroupIds.size(); ++i)
+		auto& GroupIds = Dto.GetGroupIds();
+		if (GroupIds.size() > 0)
```
Comment (MAG-MichaelK) [inline Library/src/Systems/Assets/AssetCollection.cpp:98]: Can we get empty lines before and after scope blocks, please?

File: Library/src/Systems/Assets/AssetCollection.cpp [hunk 7]
```diff
@@ -241,7 +240,11 @@ void AssetCollectionsResult::FillResultTotalCount(const csp::common::String& Jso
 	{
 		JsonDoc.Parse(JsonContent.c_str());
 
-		if (JsonDoc.HasMember("itemTotalCount"))
+		if (JsonDoc.IsArray())
```
Comment (MAG-MichaelK) [inline Library/src/Systems/Assets/AssetCollection.cpp:243]: Why was this change needed? It has worked up until now without this. Did something change in CHS?

File: Library/src/Systems/Assets/AssetSystem.cpp [hunk 8]
```diff
@@ -210,6 +210,80 @@ void AssetSystem::DeleteAssetCollection(const AssetCollection& AssetCollection,
...
+
+		INVOKE_IF_NOT_NULL(Callback, MakeInvalid<AssetCollectionsResult>());
+		return;
+	}
+
+	csp::common::String SourceSpaceId					= SourceAssetCollections[0].SpaceId;
+	std::vector<csp::common::String> AssetCollectionIds = {SourceAssetCollections[0].Id};
+
+	bool AssetCollectionsBelongToSameSpace = true;
+	for (size_t i = 1; i < SourceAssetCollections.Size(); ++i)
```
Comment (MAG-MichaelK) [inline Library/src/Systems/Assets/AssetSystem.cpp:230]: Spacing

File: Library/src/Systems/Assets/AssetSystem.cpp [hunk 9]
```diff
@@ -210,6 +210,80 @@ void AssetSystem::DeleteAssetCollection(const AssetCollection& AssetCollection,
...
+		CSP_LOG_MSG(
+			LogLevel::Error,
+			"An asset with no space ID was provided whilst attempting to perform a copy to another space. All assets must have a valid space ID.");
+
+		INVOKE_IF_NOT_NULL(Callback, MakeInvalid<AssetCollectionsResult>());
+		return;
+	}
+
+	// Verify that all source asset collections belong to the same space. If not, this qualifies as an unsupported operation.
+	if (AssetCollectionsBelongToSameSpace == false)
```
Comment (MAG-MichaelK) [inline Library/src/Systems/Assets/AssetSystem.cpp:248]: Prefer `!` over `== false`

File: Library/src/Systems/Assets/AssetSystem.cpp [hunk 10]
```diff
@@ -210,6 +210,80 @@ void AssetSystem::DeleteAssetCollection(const AssetCollection& AssetCollection,
...
+			std::nullopt,							// ParentId
+			std::nullopt,							// GroupIds
+			std::nullopt,							// Types
+			true,									// HasGroup
+			std::nullopt,							// CreatedBy
+			std::nullopt,							// PrototypeOwnerIds
+			std::nullopt,							// ReadAccessFilters
+			std::nullopt,							// WriteAccessFilters
+			SourceSpaceId,							// originalGroupId
+			DestSpaceId,							// newGroupId
```
Comment (MAG-MichaelK) [inline Library/src/Systems/Assets/AssetSystem.cpp:280]: Ah, I see we're limited by CHS in regards to group ID.

File: Library/src/Systems/Assets/AssetSystem.cpp [hunk 11]
```diff
@@ -385,14 +459,8 @@ void AssetSystem::UpdateAssetCollectionMetadata(const AssetCollection& AssetColl
 												const csp::common::Map<csp::common::String, csp::common::String>& NewMetadata,
 												AssetCollectionResultCallback Callback)
 {
-	csp::common::Optional<csp::common::String> SpaceId;
```
Comment (MAG-MichaelK) [inline Library/src/Systems/Assets/AssetSystem.cpp:388]: We need this to be optional because not all AssetCollections have a space ID. Specifically, I think space thumbnails don't.

File: Tests/CSharp/src/Tests/AssetSystemTests.cs [hunk 12]
```diff
@@ -234,7 +234,7 @@ public static void CreateAssetCollectionNoSpaceTest()
             GetAssetCollectionByName(assetSystem, testAssetCollectionName, out var assetCollection);
 
             Assert.AreEqual(assetCollection.Name, testAssetCollectionName);
-            Assert.IsTrue(assetCollection.SpaceIds.IsEmpty());
```
Comment (MAG-MichaelK) [inline Tests/CSharp/src/Tests/AssetSystemTests.cs:237]: Again, this assertion was here specifically to ensure we could create AssetCollections with no space IDs.

File: Tests/CSharp/src/Tests/AssetSystemTests.cs [hunk 13]
```diff
@@ -1380,6 +1380,110 @@ public static void thirdPartyPackagedAssetIdentifierTest()
             Assert.AreEqual(assets[0].GetThirdPartyPlatformType(), Systems.EThirdPartyPlatform.UNITY);
 
         }
+#endif
+#if RUN_ALL_UNIT_TESTS || RUN_ASSETSYSTEM_TESTS || RUN_ASSETSYSTEM_COPY_ASSET_COLLECTION_TEST
+        [Test]
+        public static void copyAssetCollectionTest()
```
Comment (MAG-MichaelK) [inline Tests/CSharp/src/Tests/AssetSystemTests.cs:1386]: Casing

File: Tests/CSharp/src/Tests/AssetSystemTests.cs [hunk 14]
```diff
@@ -1380,6 +1380,110 @@ public static void thirdPartyPackagedAssetIdentifierTest()
...
+                using var result = assetSystem.UploadAssetData(sourceAssetCollection, sourceAsset, fileSource).Result;
+                var resCode = result.GetResultCode();
+
+                Assert.AreEqual(resCode, Systems.EResultCode.Success);
+            }
+
+            // Create 'dest' space and invoke the copy
+            var destSpace = SpaceSystemTests.CreateSpace(spaceSystem, destSpaceName, testSpaceDescription, Systems.SpaceAttributes.Private, null, null, null);
+            var destAssetCollections = new Common.Array<Systems.AssetCollection>();
+            {
```
Comment (MAG-MichaelK) [inline Tests/CSharp/src/Tests/AssetSystemTests.cs:1423]: Spacing

File: Tests/CSharp/src/Tests/AssetSystemTests.cs [hunk 15]
```diff
@@ -1380,6 +1380,110 @@ public static void thirdPartyPackagedAssetIdentifierTest()
...
+                assetSystem.UploadAssetDataOnProgress += (s, e) => LogDebug($"Uploading asset... ({e.Progress}%)");
+                using var result = assetSystem.UploadAssetData(sourceAssetCollection, sourceAsset, fileSource).Result;
+                var resCode = result.GetResultCode();
+
+                Assert.AreEqual(resCode, Systems.EResultCode.Success);
+            }
+
+            // Create 'dest' space and invoke the copy
+            var destSpace = SpaceSystemTests.CreateSpace(spaceSystem, destSpaceName, testSpaceDescription, Systems.SpaceAttributes.Private, null, null, null);
+            var destAssetCollections = new Common.Array<Systems.AssetCollection>();
```
Comment (MAG-MichaelK) [inline Tests/CSharp/src/Tests/AssetSystemTests.cs:1422]: This needs to be disposed once you are done with it. It can initially be set to `null` as it looks like it's just being re-assigned later without the initial value being used anyway.

File: Tests/CSharp/src/Tests/AssetSystemTests.cs [hunk 16]
```diff
@@ -1380,6 +1380,110 @@ public static void thirdPartyPackagedAssetIdentifierTest()
...
+
+            // Validate the copied asset collection and its data
+	        {
+                Assert.AreEqual(destAssetCollections.Size(), 1UL);
+		        Assert.AreNotEqual(destAssetCollections[0].Id, sourceAssetCollection.Id);
+		        Assert.AreEqual(destAssetCollections[0].SpaceId, destSpace.Id);
+		        Assert.AreEqual(destAssetCollections[0].Type, sourceAssetCollection.Type);
+		        Assert.AreEqual(destAssetCollections[0].Tags.Size(), 1UL);
+
+                GetAssetsInCollection(assetSystem, destAssetCollections[0], out var destAssets);
```
Comment (MAG-MichaelK) [inline Tests/CSharp/src/Tests/AssetSystemTests.cs:1441]: Dispose `destAssets` when finished with it.

File: Tests/CSharp/src/Tests/AssetSystemTests.cs [hunk 17]
```diff
@@ -1380,6 +1380,110 @@ public static void thirdPartyPackagedAssetIdentifierTest()
...
+                byte[] downloadedData = new byte[result.GetDataLength()];
+                Marshal.Copy(result.GetData(), downloadedData, 0, (int)result.GetDataLength());
+
+                var fileData = File.ReadAllBytes(Path.GetFullPath("assets/test.json"));
+                Assert.IsTrue(downloadedData.SequenceEqual(fileData));
+            }
+
+            // Validating that we must have at least one asset collection to copy
+	        {
+		        Common.Array<Systems.AssetCollection> sourceAssetCollections = new Common.Array<Systems.AssetCollection>(1);
```
Comment (MAG-MichaelK) [inline Tests/CSharp/src/Tests/AssetSystemTests.cs:1457]: `Common.Array<Systems.AssetCollection> sourceAssetCollections` should be `using var sourceAssetCollections` to both ensure it gets disposed and to reduce repetition.

File: Tests/CSharp/src/Tests/AssetSystemTests.cs [hunk 18]
```diff
@@ -1380,6 +1380,110 @@ public static void thirdPartyPackagedAssetIdentifierTest()
...
+                using var result = assetSystem.CopyAssetCollectionsToSpace(sourceAssetCollections, destSpace.Id, false).Result;
+                Assert.AreEqual(result.GetResultCode(), Systems.EResultCode.Failed);
+	        }
+
+            // Validating we cannot perform a copy if the asset has no space ID
+            {
+                Common.Array<Systems.AssetCollection> assetCollections = new Common.Array<Systems.AssetCollection>(1);
```
Comment (MAG-MichaelK) [inline Tests/CSharp/src/Tests/AssetSystemTests.cs:1464]: Same here as above in regards to `using var`.

File: Tests/src/PublicAPITests/AssetSystemTests.cpp [hunk 19]
```diff
@@ -2168,4 +2162,156 @@ CSP_PUBLIC_TEST(CSPEngine, AssetSystemTests, DownloadAssetDataInvalidURLTest)
...
+		printf("Uploading source asset data...\n");
+
+		csp::common::String Uri;
+		UploadAssetData(AssetSystem, SourceAssetCollection, Asset, Source, Uri);
+	}
+
+	// Create 'dest' space and invoke the copy
+	csp::systems::Space DestSpace;
+	csp::common::Array<csp::systems::AssetCollection> DestAssetCollections;
+	{
```
Comment (MAG-MichaelK) [inline Tests/src/PublicAPITests/AssetSystemTests.cpp:2229]: Spacing

File: Tests/src/PublicAPITests/AssetSystemTests.cpp [hunk 20]
```diff
@@ -2168,4 +2162,156 @@ CSP_PUBLIC_TEST(CSPEngine, AssetSystemTests, DownloadAssetDataInvalidURLTest)
...
+		printf("Creating dest space and invoking the copy...\n");
+
+		CreateSpace(SpaceSystem, DestSpaceName, SpaceDescription, csp::systems::SpaceAttributes::Private, nullptr, nullptr, nullptr, DestSpace);
+
+		csp::common::Array<csp::systems::AssetCollection> SourceAssetCollections(&SourceAssetCollection, 1);
+		auto [Result] = AWAIT_PRE(AssetSystem, CopyAssetCollectionsToSpace, RequestPredicate, SourceAssetCollections, DestSpace.Id, false);
+
+		EXPECT_EQ(Result.GetResultCode(), csp::systems::EResultCode::Success);
+		DestAssetCollections = Result.GetAssetCollections();
```
Comment (MAG-MichaelK) [inline Tests/src/PublicAPITests/AssetSystemTests.cpp:2238]: Surround test macros with blank lines to separate them from other parts of the test.

File: Library/include/CSP/Systems/Assets/AssetCollection.h (hunk 3 above)
Comment (MAG-SamBirley) [reply Library/include/CSP/Systems/Assets/AssetCollection.h:93]: This was discussed in yesterday's engineering meeting with all client teams, CSP and CHS present. We agreed on this approach unanimously.

File: Library/include/CSP/Systems/Assets/AssetSystem.h (hunk 4 above)
Comment (MAG-SamBirley) [reply Library/include/CSP/Systems/Assets/AssetSystem.h:79]: Yep

File: Library/include/CSP/Systems/Assets/AssetSystem.h (hunk 5 above)
Comment (MAG-SamBirley) [reply Library/include/CSP/Systems/Assets/AssetSystem.h:86]: It's related to the upcoming asset library feature, which I have fleshing out in Unreal to help guide the PRD.

File: Library/src/Systems/Assets/AssetCollection.cpp (hunk 7 above)
Comment (MAG-SamBirley) [reply Library/src/Systems/Assets/AssetCollection.cpp:243]: It allows us to reuse the same object type `AssetCollectionResult` for both queries that return a single asset collection, and those that return multiple asset collections.

File: Library/src/Systems/Assets/AssetSystem.cpp (hunk 11 above)
Comment (MAG-SamBirley) [reply Library/src/Systems/Assets/AssetSystem.cpp:388]: We unilaterally agreed in the all-teams engineering meeting yesterday to represent asset collections that do not belong to a space with an empty string.

File: Tests/CSharp/src/Tests/AssetSystemTests.cs (hunk 12 above)
Comment (MAG-SamBirley) [reply Tests/CSharp/src/Tests/AssetSystemTests.cs:237]: Please see the above notes regarding the decisions made in the engineering sync

File: Library/include/CSP/Common/Array.h (hunk 2 above)
Comment (MAG-SamBirley) [reply Library/include/CSP/Common/Array.h:85]: I don't remember the specific instance that highlighted the issue here, but the issue is that if the memory being copied describes a set of non-PoD objects which hold pointers to memory that they control the lifetime over, then this constructor was leading to inadvertently shared memory lifetime management (because the pointer addresses would be directly copied to the new array's objects).

So then if one array was destroyed and the objects it described had their destructors called, and had their memory freed, then the other array's objects would be pointing at memory that was no longer valid.

To address this, this change replaces the memcpy with a per-element assignment, which makes use of the object's  assignment operator instead (making it the object's concern to safely handle memory management & object lifetimes).

File: Library/src/Systems/Assets/AssetCollection.cpp (hunk 6 above)
Comment (MAG-SamBirley) [reply Library/src/Systems/Assets/AssetCollection.cpp:99]: Addressed in latest

File: Tests/CSharp/src/Tests/AssetSystemTests.cs (hunk 16 above)
Comment (MAG-SamBirley) [reply Tests/CSharp/src/Tests/AssetSystemTests.cs:1442]: Addressed in latest

File: Tests/CSharp/src/Tests/AssetSystemTests.cs (hunk 17 above)
Comment (MAG-SamBirley) [reply Tests/CSharp/src/Tests/AssetSystemTests.cs]: Addressed in latest

File: Tests/src/PublicAPITests/AssetSystemTests.cpp (hunk 19 above)
Comment (MAG-SamBirley) [reply Tests/src/PublicAPITests/AssetSystemTests.cpp:2229]: Addressed in latest

File: Tests/src/PublicAPITests/AssetSystemTests.cpp (hunk 20 above)
Comment (MAG-SamBirley) [reply Tests/src/PublicAPITests/AssetSystemTests.cpp:2239]: Addressed in latest

File: Tests/CSharp/src/Tests/AssetSystemTests.cs (hunk 18 above)
Comment (MAG-SamBirley) [reply Tests/CSharp/src/Tests/AssetSystemTests.cs]: Addressed in latest

File: Library/include/CSP/Systems/Assets/AssetSystem.h (hunk 5 above)
Comment (MAG-MichaelK) [reply Library/include/CSP/Systems/Assets/AssetSystem.h:86]: I would prefer if we had a PRD ready to review before making changes required for said PRD.

File: Library/src/Systems/Assets/AssetCollection.cpp (hunk 7 above)
Comment (MAG-MichaelK) [reply Library/src/Systems/Assets/AssetCollection.cpp:244]: I'm not sure I follow. This change looks like 2 different ways of handling JSON that returns multiple results.

File: Library/src/Systems/Assets/AssetSystem.cpp (hunk 11 above)
Comment (MAG-MichaelK) [reply Library/src/Systems/Assets/AssetSystem.cpp:388]: Is this change supported in CHS yet, though? I believe we were having permission issues before, which is what led to us needing to remove space ID for certain asset types.

File: Library/include/CSP/Common/Array.h (hunk 2 above)
Comment (MAG-SamBirley) [reply Library/include/CSP/Common/Array.h]: Since this change isn't directly related to the overall goals of the PR, I've restored it to using memcpy. This change is no longer part of the PR, so resolving the conversation.

File: Library/include/CSP/Systems/Assets/AssetSystem.h (hunk 5 above)
Comment (MAG-SamBirley) [reply Library/include/CSP/Systems/Assets/AssetSystem.h:86]: Hey! The PRD is already in review, and is pretty clear on the requirement for asset duplication, so this is definitely required.

File: Library/include/CSP/Systems/Assets/AssetCollection.h (hunk 3 above)
Comment (MAG-SamBirley) [reply Library/include/CSP/Systems/Assets/AssetCollection.h:93]: Understood, however since the group consensus was to adopt this approach, and our services team are also considering doing similarly at their API level, I don't see any harm in including this part in the PR. It helps keeps the new `Copy` function cleaner too!

**Review by MAG-AdamThorn — APPROVED** [review]

File: Tests/CSharp/src/Tests/AssetSystemTests.cs (hunk 15 above)
Comment (MAG-SamBirley) [reply Tests/CSharp/src/Tests/AssetSystemTests.cs]: Resolved!

File: Tests/CSharp/src/Tests/AssetSystemTests.cs (hunk 14 above)
Comment (MAG-SamBirley) [reply Tests/CSharp/src/Tests/AssetSystemTests.cs:1424]: Resolved!

File: Tests/CSharp/src/Tests/AssetSystemTests.cs (hunk 13 above)
Comment (MAG-SamBirley) [reply Tests/CSharp/src/Tests/AssetSystemTests.cs]: Resolved!

File: Library/src/Systems/Assets/AssetSystem.cpp (hunk 9 above)
Comment (MAG-SamBirley) [reply Library/src/Systems/Assets/AssetSystem.cpp]: Resolved!

File: Library/src/Systems/Assets/AssetSystem.cpp (hunk 8 above)
Comment (MAG-SamBirley) [reply Library/src/Systems/Assets/AssetSystem.cpp:231]: Resolved!

File: Library/src/Systems/Assets/AssetCollection.cpp (hunk 7 above)
Comment (MAG-SamBirley) [reply Library/src/Systems/Assets/AssetCollection.cpp:244]: That's right. This function can now handle two different payloads, one which describes a single asset collection, and one that can return many.
